[close #59] Use relative path for cache keys

schneems opened the pull request:
This fix is targeted at 3.x. This problem may or may not exist on 4.x I've been unable to get 4.x working with a Rails app.

Right now the cache is using absolute paths to generate cache keys, this doesn't work if you end up running `rake assets:precompile` in different directories. Say you've moved your project or are running on a different server with a different absolute path. Or say, on every build on Heroku ever. This was reported in #59.

This change introduces UnloadedAsset class that is responsible for among other things generating consistent cache keys.

Previously when running https://gist.github.com/schneems/85f592ba2773761dfcf3 the output would show that the second build took over 38 seconds to build. With this patch applied it takes around 3.38 seconds.

Most of this commit is docs, as the current behavior and use of the cache is not really documented, and extremely difficult to follow in the code. All existing behavior is preserved, all tests pass, an extra test was added to ensure that a project built using a previously built directory will be fast.

elia commented:
I can be wrong of course but iirc sprockets has a number of ways to require files local to a specific root regardless of the paths sequence and the possible overrides, that's probably why absolute paths were chosen in the first place. For example I'm quite sure `require_tree` has this behavior. That means that one relative path can be used from different locations. 

Not sure how to solve this tho. 😕

schneems commented:
Locally I've got all the tests to pass minus 7 in https://github.com/rails/sprockets/blob/master/test/test_performance.rb which i'm working on. Doesn't look like using relative paths broke any logic.

schneems commented:
ALL TESTS PASS  🐘🚀💘🍰! I think some of the behavior you're referring to is covered since we fall back to an absolute path if the path is not inside of `root`. Still need some time for cleanup and documentation, and want to add at least one test for this behavior so we don't get a regression

schneems commented:
I'm done with cleanup, added docs and a test for current behavior. I also updated the PR description. I tried to reproduce the problem with Sprockets 4.x and couldn't so it looks like this issue is isolated to only 3.x though I have no idea why and would like others to double check me  to see if they can reproduce the original issue with sprockets 4.x.

schneems commented:
I take it back this issue may exist on 4.x. I was getting fast times because 4.x wasn't producing ANY assets, check out https://github.com/schneems/test-move-sprockets-4.x would appreciate your help with repro.

mferrier commented:
:+1: :+1: :+1:

elia commented:
Here's an easy way to reproduce the issue for 4.x (master) https://github.com/rails/sprockets/pull/86

schneems commented:
@elia thanks, that did the trick. I was able to repro problem with Rails 4: https://github.com/rails/sprockets/issues/90

rafaelfranca commented on lib/sprockets/loader.rb:
```diff
@@ -12,37 +12,152 @@
+    # The +uri+ is a complete uri to a file including schema
+    # such as: "file:///Full/path/app/assets/javascripts/application.js?type=application/javascript"
+    #
+    # The +env+ contains the current "environment" which means it has access to every gosh-dang method
```
Can we keep using TomDoc here? http://tomdoc.org/

rafaelfranca commented on lib/sprockets/loader.rb:
```diff
@@ -133,18 +249,52 @@ def load_asset_by_uri(uri, filename, params)
         }.compact.max
-        asset[:mtime] ||= self.stat(filename).mtime.to_i
+        asset[:mtime] ||= self.stat(unloaded.filename).mtime.to_i
+
```
:scissors:

rafaelfranca commented on test/test_performance.rb:
```diff
@@ -106,6 +106,34 @@ def teardown
+    assert_no_redundant_bundle_processor_calls
+    assert_no_redundant_cache_set_calls
+
+
```
:scissors:

rafaelfranca commented:
The general idea seems good. We need to keep the documentation style of the project though.

rafaelfranca commented on lib/sprockets/loader.rb:
```diff
@@ -12,37 +12,198 @@
+
+    # Internal: Full file path without schema
+    #
+    # This returns a string containging the full path to the asset without the schema.
```
`containing`

rafaelfranca commented on lib/sprockets/loader.rb:
```diff
@@ -12,37 +12,198 @@
+    # for a given asset path and filename digest.
+    #
+    # A dependency can refer to either an asset i.e. index.js
+    # may rely on jquery.js (so jquery.js is a depndency), or other factors that may affect
```
`dependency`

rafaelfranca commented:
:shipit:

schneems commented:
Thanks for the review, travis is green. Merging in!
